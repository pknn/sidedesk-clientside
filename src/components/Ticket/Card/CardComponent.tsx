import React from 'react'
import { DraggableProvided, DraggableStateSnapshot } from 'react-beautiful-dnd'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faUser } from '@fortawesome/free-solid-svg-icons'

interface DataProps {
  id: number
  title: string
  reporterName: string
}

interface DraggableProps {
  draggableProvided: DraggableProvided
  draggableStateSnapshot: DraggableStateSnapshot
}

type ComponentProps = DataProps & DraggableProps

const getCardClassNames = (isDragging: boolean) => {
  return [
    'mb-2 w-full hover:bg-gray-100 p-4 rounded shadow cursor-pointer',
    isDragging ? 'bg-gray-200' : 'bg-white',
  ].join(' ')
}

export const CardComponent = ({
  id,
  title,
  reporterName,
  draggableProvided,
  draggableStateSnapshot,
}: ComponentProps): JSX.Element => (
  <div
    ref={draggableProvided.innerRef}
    {...draggableProvided.draggableProps}
    {...draggableProvided.dragHandleProps}
    className={getCardClassNames(draggableStateSnapshot.isDragging)}
  >
    <div className="text-xs text-gray-400 flex items-center">#{id}</div>
    <div className="flex-1 text-sm">{title}</div>
    <div className="mt-2 flex items-baseline">
      <FontAwesomeIcon className="mr-1" size="xs" icon={faUser} />
      <span className="align-baseline text-xs">{reporterName}</span>
    </div>
  </div>
)
