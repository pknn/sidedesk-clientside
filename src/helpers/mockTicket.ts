import faker from 'faker'

import { Ticket, TicketForm, TicketStatus } from 'app/types/Ticket'

export const ticketStatusOptions: TicketStatus[] = [
  TicketStatus.Pending,
  TicketStatus.Accepted,
  TicketStatus.Resolved,
  TicketStatus.Rejected,
]

type DateIndex = 0 | 1 | 2

const dates: Date[] = [
  new Date(1623862800),
  new Date(1623949200),
  new Date(1624035600),
]

const getFakeDate = (dateIndex: DateIndex): Date => dates[dateIndex]

const getFakeTicket = (
  status?: TicketStatus,
  dateIndex?: DateIndex,
): Ticket => {
  const firstName = faker.name.firstName()
  const lastName = faker.name.lastName()
  const fullName = `${firstName} ${lastName}`
  return {
    id: faker.datatype.number(),
    title: faker.lorem.sentence(10),
    description: faker.lorem.sentences(4),
    reporterName: fullName,
    reporterEmail: faker.internet.email(firstName, lastName),
    status: status || faker.random.arrayElement(ticketStatusOptions),
    createdAt: getFakeDate(dateIndex || 0),
    updatedAt: getFakeDate(dateIndex || 0),
  }
}

export const getMockTicket = (status?: TicketStatus): Ticket => {
  faker.seed(96421)
  return getFakeTicket(status)
}

export const getMockTickets = (n: number, status?: TicketStatus): Ticket[] => {
  faker.seed(96421)
  const tickets: Ticket[] = []
  for (let i = 0; i < n; i += 1) {
    const fa = i % 3 == 0
    const fb = i % 3 == 1
    tickets.push(getFakeTicket(status, fa ? 0 : fb ? 1 : 2))
  }
  return tickets
}

export const getTickets = (n: number, status: TicketStatus): Ticket[] => {
  const tickets: Ticket[] = []
  for (let i = 0; i < n; i += 1) {
    const fa = i % 3 == 0
    const fb = i % 3 == 1
    tickets.push(getFakeTicket(status, fa ? 0 : fb ? 1 : 2))
  }
  return tickets
}
