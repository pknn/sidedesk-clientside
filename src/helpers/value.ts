export const getOr =
  <T>(defaultVal: T) =>
  (val: T | undefined): T =>
    val || defaultVal

export const getOrElse = <T>(val: T | undefined, defaultVal: T): T =>
  getOr(defaultVal)(val)
