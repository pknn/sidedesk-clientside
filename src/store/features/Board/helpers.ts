import { DraggableLocation } from 'react-beautiful-dnd'

import {
  MoveCrossLaneActionPayload,
  MoveInLaneActionPayload,
  TicketPosition,
} from 'app/store/features/Board/types'
import { Ticket } from 'app/types/Ticket'
import {
  getStatusFromKey,
  getStatusFromString,
  TicketStatusKeys,
} from 'app/helpers/statusMappers'

export const getMoveInLaneActionPayload = (
  source: DraggableLocation,
  destination: DraggableLocation,
): MoveInLaneActionPayload => ({
  laneId: source.droppableId,
  from: source.index,
  to: destination.index,
})

export const getMoveCrossLaneActionPayload = (
  source: DraggableLocation,
  destination: DraggableLocation,
): MoveCrossLaneActionPayload => ({
  from: {
    laneId: source.droppableId,
    index: source.index,
  },
  to: {
    laneId: destination.droppableId,
    index: destination.index,
  },
})

export const getReorderedTicketList = (
  tickets: Ticket[],
  from: number,
  to: number,
) => {
  const result = [...tickets]
  const [movingTicket] = result.splice(from, 1)
  result.splice(to, 0, movingTicket)

  return result
}

export const getMovedTicketLists = (
  source: Ticket[],
  sink: Ticket[],
  from: number,
  to: TicketPosition,
): [Ticket[], Ticket[]] => [
  source.filter((_, index) => index !== from),
  sink
    .slice(0, to.index)
    .concat({
      ...source[from],
      status: getStatusFromString(to.laneId as TicketStatusKeys),
    })
    .concat(sink.slice(to.index)),
]
