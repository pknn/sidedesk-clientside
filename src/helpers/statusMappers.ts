import { BoardState } from 'app/types/Board'
import { TicketStatus } from 'app/types/Ticket'

export type BoardKeys = keyof BoardState
export type TicketStatusKeys = 'Pending' | 'Accepted' | 'Resolved' | 'Rejected'

const statusKeyMap: Record<TicketStatus, BoardKeys> = {
  [TicketStatus.Pending]: 'pendingTickets',
  [TicketStatus.Accepted]: 'acceptedTickets',
  [TicketStatus.Resolved]: 'resolvedTickets',
  [TicketStatus.Rejected]: 'rejectedTickets',
}

const keyStatusMap: Record<BoardKeys, TicketStatus> = (() => {
  const entries = Object.entries(statusKeyMap)
  const reversedMappedEntries = entries.map(([k, v]) => [v, k])
  return Object.fromEntries(reversedMappedEntries) as Record<
    BoardKeys,
    TicketStatus
  >
})()

const stringStatusMap: Record<TicketStatusKeys, TicketStatus> = {
  Pending: TicketStatus.Pending,
  Accepted: TicketStatus.Accepted,
  Resolved: TicketStatus.Resolved,
  Rejected: TicketStatus.Rejected,
}

export const getKeyFromStatus = (status: TicketStatus): BoardKeys =>
  statusKeyMap[status]

export const getStatusFromKey = (key: BoardKeys): TicketStatus =>
  keyStatusMap[key]

export const getStatusFromString = (key: TicketStatusKeys): TicketStatus =>
  stringStatusMap[key]

export const getKeyFromString = (key: string): BoardKeys =>
  getKeyFromStatus(stringStatusMap[key as TicketStatusKeys])
