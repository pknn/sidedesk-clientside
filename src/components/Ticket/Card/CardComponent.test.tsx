import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { CardComponent } from './CardComponent'
import { getMockTicket } from 'app/helpers/mockTicket'

describe('<CardComponent />', () => {
  it('should render correctly', () => {
    const mockTicket = getMockTicket()
    const shallowMountedComponent = shallow(
      <CardComponent
        id={mockTicket.id}
        title={mockTicket.title}
        reporterName={mockTicket.reporterName}
      />,
    )

    expect(toJson(shallowMountedComponent)).toMatchSnapshot()
  })
})
