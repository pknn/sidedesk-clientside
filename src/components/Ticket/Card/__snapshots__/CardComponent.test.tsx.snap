// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardComponent /> should render correctly 1`] = `
<div
  className="mb-2 w-full bg-white hover:bg-gray-100 p-4 rounded shadow cursor-pointer"
>
  <div
    className="text-xs text-gray-400 flex items-center"
  >
    #
    7532
  </div>
  <div
    className="flex-1 text-sm"
  >
    Repudiandae accusantium et nihil aut in animi numquam quis laborum.
  </div>
  <div
    className="mt-2 flex items-baseline"
  >
    <FontAwesomeIcon
      border={false}
      className="mr-1"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f007",
            "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
          ],
          "iconName": "user",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size="xs"
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <span
      className="align-baseline text-xs"
    >
      Jalen Rutherford
    </span>
  </div>
</div>
`;
